tests/queries/0_stateless/02007_test_any_all_operators.sql

```diff
@@ -12,7 +12,7 @@ select 1 != any (select 1 from numbers(10));
 select 1 != any (select number from numbers(10));
 
 select number as a from numbers(10) where a == any (select number from numbers(3, 3));
-select number as a from numbers(10) where a != any (select 5 from numbers(3, 3));
+select number as a from numbers(10) where a != any (select 5 from numbers(3, 3)) settings use_index_for_in_with_subqueries=0;
```

Comment:
For me this query fails without disabling the index for in with subqueries. Could you please check why? I don't think that is a good behavior.

Reply:
I think there is a bug which is not related with this PR. For CI checking I just try to avoid it. Or I just leave it?

Reply:
So when I run this query the printed stack trace is the following:
```
2023.06.30 13:32:31.131687 [ 3165400 ] {66a03b7d-8368-4577-b03b-76a242885d55} <Error> TCPHandler: Code: 53. DB::Exception: Types of column 1 in section IN don't match: UInt64 on the left, UInt8 on the right. (TYPE_MISMATCH), Stack trace (when copying this message, always include the lines below):

0. /home/ubuntu/Repos/ClickHouse/contrib/llvm-project/libcxx/include/exception:134: std::exception::capture() @ 0x000000001adcbb22 in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
1. /home/ubuntu/Repos/ClickHouse/contrib/llvm-project/libcxx/include/exception:112: std::exception::exception[abi:v15000]() @ 0x000000001adcbaed in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
2. /home/ubuntu/Repos/ClickHouse/base/poco/Foundation/src/Exception.cpp:27: Poco::Exception::Exception(String const&, int) @ 0x00000000341c1460 in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
3. /home/ubuntu/Repos/ClickHouse/src/Common/Exception.cpp:90: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000024368dae in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
4. /home/ubuntu/Repos/ClickHouse/src/Common/Exception.h:54: DB::Exception::Exception(String&&, int, bool) @ 0x000000001adbfc4a in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
5. /home/ubuntu/Repos/ClickHouse/src/Common/Exception.h:81: DB::Exception::Exception<String, String, String>(int, FormatStringHelperImpl<std::type_identity<String>::type, std::type_identity<String>::type, std::type_identity<String>::type>, String&&, String&&, String&&) @ 0x000000001ae666a8 in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
6. /home/ubuntu/Repos/ClickHouse/src/Interpreters/Set.cpp:460: DB::Set::checkTypesEqual(unsigned long, std::shared_ptr<DB::IDataType const> const&) const @ 0x000000002d39a63a in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
7. /home/ubuntu/Repos/ClickHouse/src/Storages/MergeTree/KeyCondition.cpp:1647: DB::KeyCondition::tryPrepareSetIndex(DB::RPNBuilderFunctionTreeNode const&, DB::KeyCondition::RPNElement&, unsigned long&) @ 0x000000002e34246d in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
8. /home/ubuntu/Repos/ClickHouse/src/Storages/MergeTree/KeyCondition.cpp:1918: DB::KeyCondition::extractAtomFromTree(DB::RPNBuilderTreeNode const&, DB::KeyCondition::RPNElement&) @ 0x000000002e3439dc in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
9. /home/ubuntu/Repos/ClickHouse/src/Storages/MergeTree/KeyCondition.cpp:1165: DB::KeyCondition::KeyCondition(std::shared_ptr<DB::IAST> const&, absl::lts_20211102::InlinedVector<std::shared_ptr<DB::IAST>, 7ul, std::allocator<std::shared_ptr<DB::IAST>>> const&, DB::Block, std::shared_ptr<DB::PreparedSets>, std::shared_ptr<DB::Context const>, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::ExpressionActions> const&, std::unordered_set<String, std::hash<String>, std::equal_to<String>, std::allocator<String>>, bool, bool)::$_0::operator()(DB::RPNBuilderTreeNode const&, DB::KeyCondition::RPNElement&) const @ 0x000000002e352f48 in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
10. /home/ubuntu/Repos/ClickHouse/contrib/llvm-project/libcxx/include/__functional/invoke.h:394: decltype(std::declval<DB::KeyCondition::KeyCondition(std::shared_ptr<DB::IAST> const&, absl::lts_20211102::InlinedVector<std::shared_ptr<DB::IAST>, 7ul, std::allocator<std::shared_ptr<DB::IAST>>> const&, DB::Block, std::shared_ptr<DB::PreparedSets>, std::shared_ptr<DB::Context const>, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::ExpressionActions> const&, std::unordered_set<String, std::hash<String>, std::equal_to<String>, std::allocator<String>>, bool, bool)::$_0&>()(std::declval<DB::RPNBuilderTreeNode const&>(), std::declval<DB::KeyCondition::RPNElement&>())) std::__invoke[abi:v15000]<DB::KeyCondition::KeyCondition(std::shared_ptr<DB::IAST> const&, absl::lts_20211102::InlinedVector<std::shared_ptr<DB::IAST>, 7ul, std::allocator<std::shared_ptr<DB::IAST>>> const&, DB::Block, std::shared_ptr<DB::PreparedSets>, std::shared_ptr<DB::Context const>, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::ExpressionActions> const&, std::unordered_set<String, std::hash<String>, std::equal_to<String>, std::allocator<String>>, bool, bool)::$_0&, DB::RPNBuilderTreeNode const&, DB::KeyCondition::RPNElement&>(DB::KeyCondition::KeyCondition(std::shared_ptr<DB::IAST> const&, absl::lts_20211102::InlinedVector<std::shared_ptr<DB::IAST>, 7ul, std::allocator<std::shared_ptr<DB::IAST>>> const&, DB::Block, std::shared_ptr<DB::PreparedSets>, std::shared_ptr<DB::Context const>, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::ExpressionActions> const&, std::unordered_set<String, std::hash<String>, std::equal_to<String>, std::allocator<String>>, bool, bool)::$_0&, DB::RPNBuilderTreeNode const&, DB::KeyCondition::RPNElement&) @ 0x000000002e352f05 in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
11. /home/ubuntu/Repos/ClickHouse/contrib/llvm-project/libcxx/include/__functional/invoke.h:470: bool std::__invoke_void_return_wrapper<bool, false>::__call<DB::KeyCondition::KeyCondition(std::shared_ptr<DB::IAST> const&, absl::lts_20211102::InlinedVector<std::shared_ptr<DB::IAST>, 7ul, std::allocator<std::shared_ptr<DB::IAST>>> const&, DB::Block, std::shared_ptr<DB::PreparedSets>, std::shared_ptr<DB::Context const>, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::ExpressionActions> const&, std::unordered_set<String, std::hash<String>, std::equal_to<String>, std::allocator<String>>, bool, bool)::$_0&, DB::RPNBuilderTreeNode const&, DB::KeyCondition::RPNElement&>(DB::KeyCondition::KeyCondition(std::shared_ptr<DB::IAST> const&, absl::lts_20211102::InlinedVector<std::shared_ptr<DB::IAST>, 7ul, std::allocator<std::shared_ptr<DB::IAST>>> const&, DB::Block, std::shared_ptr<DB::PreparedSets>, std::shared_ptr<DB::Context const>, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::ExpressionActions> const&, std::unordered_set<String, std::hash<String>, std::equal_to<String>, std::allocator<String>>, bool, bool)::$_0&, DB::RPNBuilderTreeNode const&, DB::KeyCondition::RPNElement&) @ 0x000000002e352ec5 in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
12. /home/ubuntu/Repos/ClickHouse/contrib/llvm-project/libcxx/include/__functional/function.h:235: std::__function::__default_alloc_func<DB::KeyCondition::KeyCondition(std::shared_ptr<DB::IAST> const&, absl::lts_20211102::InlinedVector<std::shared_ptr<DB::IAST>, 7ul, std::allocator<std::shared_ptr<DB::IAST>>> const&, DB::Block, std::shared_ptr<DB::PreparedSets>, std::shared_ptr<DB::Context const>, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::ExpressionActions> const&, std::unordered_set<String, std::hash<String>, std::equal_to<String>, std::allocator<String>>, bool, bool)::$_0, bool (DB::RPNBuilderTreeNode const&, DB::KeyCondition::RPNElement&)>::operator()[abi:v15000](DB::RPNBuilderTreeNode const&, DB::KeyCondition::RPNElement&) @ 0x000000002e352e85 in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
13. /home/ubuntu/Repos/ClickHouse/contrib/llvm-project/libcxx/include/__functional/function.h:716: bool std::__function::__policy_invoker<bool (DB::RPNBuilderTreeNode const&, DB::KeyCondition::RPNElement&)>::__call_impl<std::__function::__default_alloc_func<DB::KeyCondition::KeyCondition(std::shared_ptr<DB::IAST> const&, absl::lts_20211102::InlinedVector<std::shared_ptr<DB::IAST>, 7ul, std::allocator<std::shared_ptr<DB::IAST>>> const&, DB::Block, std::shared_ptr<DB::PreparedSets>, std::shared_ptr<DB::Context const>, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::ExpressionActions> const&, std::unordered_set<String, std::hash<String>, std::equal_to<String>, std::allocator<String>>, bool, bool)::$_0, bool (DB::RPNBuilderTreeNode const&, DB::KeyCondition::RPNElement&)>>(std::__function::__policy_storage const*, DB::RPNBuilderTreeNode const&, DB::KeyCondition::RPNElement&) @ 0x000000002e352e4d in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
14. /home/ubuntu/Repos/ClickHouse/contrib/llvm-project/libcxx/include/__functional/function.h:848: std::__function::__policy_func<bool (DB::RPNBuilderTreeNode const&, DB::KeyCondition::RPNElement&)>::operator()[abi:v15000](DB::RPNBuilderTreeNode const&, DB::KeyCondition::RPNElement&) const @ 0x000000002e381ce6 in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
15. /home/ubuntu/Repos/ClickHouse/contrib/llvm-project/libcxx/include/__functional/function.h:1187: std::function<bool (DB::RPNBuilderTreeNode const&, DB::KeyCondition::RPNElement&)>::operator()(DB::RPNBuilderTreeNode const&, DB::KeyCondition::RPNElement&) const @ 0x000000002e381b25 in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
16. /home/ubuntu/Repos/ClickHouse/src/Storages/MergeTree/RPNBuilder.h:250: DB::RPNBuilder<DB::KeyCondition::RPNElement>::traverseTree(DB::RPNBuilderTreeNode const&) @ 0x000000002e381838 in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
17. /home/ubuntu/Repos/ClickHouse/src/Storages/MergeTree/RPNBuilder.h:216: DB::RPNBuilder<DB::KeyCondition::RPNElement>::RPNBuilder(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const>, DB::Block, std::shared_ptr<DB::PreparedSets>, std::function<bool (DB::RPNBuilderTreeNode const&, DB::KeyCondition::RPNElement&)> const&) @ 0x000000002e36bc27 in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
18. /home/ubuntu/Repos/ClickHouse/src/Storages/MergeTree/KeyCondition.cpp:1160: DB::KeyCondition::KeyCondition(std::shared_ptr<DB::IAST> const&, absl::lts_20211102::InlinedVector<std::shared_ptr<DB::IAST>, 7ul, std::allocator<std::shared_ptr<DB::IAST>>> const&, DB::Block, std::shared_ptr<DB::PreparedSets>, std::shared_ptr<DB::Context const>, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::ExpressionActions> const&, std::unordered_set<String, std::hash<String>, std::equal_to<String>, std::allocator<String>>, bool, bool) @ 0x000000002e33e849 in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
19. /home/ubuntu/Repos/ClickHouse/src/Storages/MergeTree/KeyCondition.cpp:1176: DB::KeyCondition::KeyCondition(DB::SelectQueryInfo const&, std::shared_ptr<DB::Context const>, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::ExpressionActions> const&, bool, bool) @ 0x000000002e33ed52 in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
20. /home/ubuntu/Repos/ClickHouse/src/Storages/System/StorageSystemNumbers.cpp:218: DB::StorageSystemNumbers::read(std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::StorageSnapshot> const&, DB::SelectQueryInfo&, std::shared_ptr<DB::Context const>, DB::QueryProcessingStage::Enum, unsigned long, unsigned long) @ 0x000000002aeb6ecb in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
21. /home/ubuntu/Repos/ClickHouse/src/Storages/IStorage.cpp:135: DB::IStorage::read(DB::QueryPlan&, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::StorageSnapshot> const&, DB::SelectQueryInfo&, std::shared_ptr<DB::Context const>, DB::QueryProcessingStage::Enum, unsigned long, unsigned long) @ 0x000000002db4ea21 in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
22. /home/ubuntu/Repos/ClickHouse/src/Interpreters/InterpreterSelectQuery.cpp:2514: DB::InterpreterSelectQuery::executeFetchColumns(DB::QueryProcessingStage::Enum, DB::QueryPlan&) @ 0x000000002d071f8f in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
23. /home/ubuntu/Repos/ClickHouse/src/Interpreters/InterpreterSelectQuery.cpp:0: DB::InterpreterSelectQuery::executeImpl(DB::QueryPlan&, std::optional<DB::Pipe>) @ 0x000000002d069b5e in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
24. /home/ubuntu/Repos/ClickHouse/src/Interpreters/InterpreterSelectQuery.cpp:899: DB::InterpreterSelectQuery::buildQueryPlan(DB::QueryPlan&) @ 0x000000002d068cbd in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
25. /home/ubuntu/Repos/ClickHouse/src/Interpreters/InterpreterSelectWithUnionQuery.cpp:303: DB::InterpreterSelectWithUnionQuery::buildQueryPlan(DB::QueryPlan&) @ 0x000000002d159094 in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
26. /home/ubuntu/Repos/ClickHouse/src/Interpreters/InterpreterSelectWithUnionQuery.cpp:377: DB::InterpreterSelectWithUnionQuery::execute() @ 0x000000002d159be0 in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
27. /home/ubuntu/Repos/ClickHouse/src/Interpreters/executeQuery.cpp:746: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x000000002d6203f7 in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
28. /home/ubuntu/Repos/ClickHouse/src/Interpreters/executeQuery.cpp:1168: DB::executeQuery(String const&, std::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) @ 0x000000002d61b944 in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
29. /home/ubuntu/Repos/ClickHouse/src/Server/TCPHandler.cpp:421: DB::TCPHandler::runImpl() @ 0x000000002eaf1a4d in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
30. /home/ubuntu/Repos/ClickHouse/src/Server/TCPHandler.cpp:2057: DB::TCPHandler::run() @ 0x000000002eb03f25 in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
31. /home/ubuntu/Repos/ClickHouse/base/poco/Net/src/TCPServerConnection.cpp:43: Poco::Net::TCPServerConnection::start() @ 0x0000000033ffd4d9 in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
```

In position 20 there is the following line:
```
20. /home/ubuntu/Repos/ClickHouse/src/Storages/System/StorageSystemNumbers.cpp:218: DB::StorageSystemNumbers::read(std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::StorageSnapshot> const&, DB::SelectQueryInfo&, std::shared_ptr<DB::Context const>, DB::QueryProcessingStage::Enum, unsigned long, unsigned long) @ 0x000000002aeb6ecb in /home/ubuntu/Repos/ClickHouse/build/programs/clickhouse
```

I think this is connected to your changes. Furthermore, the tests are not failing on master without the extra setting, but failing on your bracnh. So maybe not your changes caused the issue, but you changes definitely trigger the issue in cases where it wasn't an issue before. Either way, I think it would be nice to fix this before merging.

Reply:
Ok, I will fix it next Monday.

Reply:
@antaljanosbenjamin fixed, please help review it.

Reply:
I think it is totally valid solution for the issue, thanks for the fix!